File: day15_part1_2020.rb

input = File.read('input.txt').split(',').map(&:to_i)

last_spoken = {}
input.each_with_index { |num, idx| last_spoken[num] = idx + 1 }

turn = input.length + 1
prev_num = input.last

while turn <= 2020
  if last_spoken[prev_num].nil?
    spoken = 0
  else
    spoken = turn - 1 - last_spoken[prev_num]
  end

  last_spoken[prev_num] = turn - 1
  prev_num = spoken
  turn += 1
end

puts prev_num

File: clojure/day15_part1_2020.clj

(defn play-game [starting-numbers n]
  (let [spoken-numbers (vec starting-numbers)
        spoken-map (zipmap starting-numbers (range 1 (count starting-numbers)))
        last-spoken (peek starting-numbers)]
    (loop [turn (count starting-numbers)
           last-spoken last-spoken
           spoken-numbers spoken-numbers
           spoken-map spoken-map]
      (if (= turn n)
        last-spoken
        (let [next-spoken (if (contains? spoken-map last-spoken)
                            (- turn (get spoken-map last-spoken))
                            0)]
          (recur (inc turn)
                 next-spoken
                 (conj spoken-numbers next-spoken)
                 (assoc spoken-map last-spoken turn)))))))

(def input (slurp "input.txt"))
(def starting-numbers (mapv #(Integer/parseInt %) (clojure.string/split input #",")))

(println (play-game starting-numbers 2020))
